keys.go

```diff
@@ -50,116 +74,199 @@ func findKey(session *pkcs11Session, id []byte, label []byte, keyclass *uint, ke
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, 0),
+	}
+	if attributes, err = session.ctx.GetAttributeValue(session.handle, *privHandle, attributes); err != nil {
+		return nil, err
+	}
+	id := attributes[0].Value
+	label := attributes[1].Value
+	keyType := bytesToUlong(attributes[2].Value)
+
+	// Find the public half which has a matching CKA_ID
+	pubHandle, err := findKey(session, id, label, uintPtr(pkcs11.CKO_PUBLIC_KEY), &keyType)
+	if err != nil {
+		return nil, err
+	}
+	if pubHandle == nil {
+		// We can't return a Signer if we don't have private and public key. Treat it as an error.
+		return nil, errors.New("could not find public key to match private key")
+	}
+
+	var pub crypto.PublicKey
+	switch keyType {
+	case pkcs11.CKK_DSA:
+		if pub, err = exportDSAPublicKey(session, *pubHandle); err != nil {
+			return nil, err
+		}
+		return &pkcs11PrivateKeyDSA{
+			pkcs11PrivateKey: pkcs11PrivateKey{
+				pkcs11Object: pkcs11Object{
+					handle:  *privHandle,
+					context: c,
+				},
+				pubKeyHandle: *pubHandle,
+				pubKey:       pub,
+			}}, nil
+
+	case pkcs11.CKK_RSA:
+		if pub, err = exportRSAPublicKey(session, *pubHandle); err != nil {
+			return nil, err
+		}
+		return &pkcs11PrivateKeyRSA{
+			pkcs11PrivateKey: pkcs11PrivateKey{
+				pkcs11Object: pkcs11Object{
+					handle:  *privHandle,
+					context: c,
+				},
+				pubKeyHandle: *pubHandle,
+				pubKey:       pub,
+			}}, nil
+
+	case pkcs11.CKK_ECDSA:
+		if pub, err = exportECDSAPublicKey(session, *pubHandle); err != nil {
+			return nil, err
+		}
+		return &pkcs11PrivateKeyECDSA{
+			pkcs11PrivateKey: pkcs11PrivateKey{
+				pkcs11Object: pkcs11Object{
+					handle:  *privHandle,
+					context: c,
+				},
+				pubKeyHandle: *pubHandle,
+				pubKey:       pub,
+			}}, nil
+
+	default:
+		return nil, errors.Errorf("unsupported key type: %X", keyType)
+	}
+}
+
 // FindKeyPair retrieves a previously created asymmetric key pair, or nil if it cannot be found.
 //
 // At least one of id and label must be specified. If the private key is found, but the public key is
 // not, an error is returned because we cannot implement crypto.Signer without the public key.
 func (c *Context) FindKeyPair(id []byte, label []byte) (Signer, error) {
+	result, err := c.FindKeyPairs(id, label)
+	if err != nil {
+		return nil, err
+	}
 
-	if c.closed.Get() {
-		return nil, errClosed
+	if len(result) == 0 {
+		return nil, nil
 	}
 
-	var k Signer
+	return result[0], nil
+}
 
-	err := c.withSession(func(session *pkcs11Session) error {
+// FindKeyPairs retrieves all matching asymmetric key pairs, or a nil slice if none can be found.
+//
+// At least one of id and label must be specified.
+// If a private key is found, but the corresponding public key is not, the key is not returned because we cannot
+// implement crypto.Signer without the public key.
+func (c *Context) FindKeyPairs(id []byte, label []byte) ([]Signer, error) {
+	if id == nil && label == nil {
```

Comment:
Ok, I omitted it from some of the functions because they call through to other functions which do the check, but I'll add it back at the beginning of each.